### PR TITLE
Refactoring database handling to allow for unit testing

### PR DESCRIPTION
**Pull request by marcemmers**
## Describe your changes

Refactored and improved the database classes:

- Introduced new `DatabaseConnection` which is a pure SQL interface with helper functions but without implementation details.
- `DatabaseHandlerBase` is renamed into `DatabaseHandlerCommon` and contains the common functionality of v16 and v201 to be used by the `MessageQueue` class.
- `SQLiteStatement` still exists but can now only be made from a `DatabaseConnection` removing all the exposed raw C pointers that were still being used there.
- `DatabaseConnection` and `SQLiteStatement` are used through interfaces making it possible to mock them if so desired.

## Issue ticket number and link

https://github.com/EVerest/libocpp/issues/490

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

File: include/ocpp/common/database/database_handler_common.hpp [hunk 1]
```diff
@@ -20,9 +20,9 @@ struct DBTransactionMessage {
     std::string unique_id;
 };
 
-class DatabaseHandlerBase {
+class DatabaseHandlerCommon {
 protected:
-    sqlite3* db;
+    std::shared_ptr<DatabaseConnectionInterface> database;
```
Comment (Pietfried) [inline include/ocpp/common/database/database_handler_common.hpp:25]: This looks like it can be a unique_ptr, only owned by the DatabaseHandlerCommon

File: include/ocpp/v201/database_handler.hpp [hunk 2]
```diff
@@ -31,21 +31,18 @@ class DatabaseHandler : public ocpp::common::DatabaseHandlerBase {
...
     void insert_availability(int32_t evse_id, int32_t connector_id, OperationalStatusEnum operational_status,
                              bool replace);
     OperationalStatusEnum get_availability(int32_t evse_id, int32_t connector_id);
 
 public:
-    DatabaseHandler(const fs::path& database_path, const fs::path& sql_init_path);
+    DatabaseHandler(std::shared_ptr<common::DatabaseConnectionInterface> database, const fs::path& sql_init_path);
 
     /// \brief Opens connection to database file
     void open_connection();
```
Comment (Pietfried) [inline include/ocpp/v201/database_handler.hpp:44]: `open_connection` and `close_connection` could be moved to the `DatabaseHandlerCommon`. This requires some additional initialization adjustments because of the `nr_of_connectors` argument for `open_connection` in v16. We could also do this in another PR. What do you think?

File: include/ocpp/v16/database_handler.hpp [hunk 3]
```diff
@@ -38,24 +38,23 @@ struct TransactionEntry {
 };
 
 /// \brief This class handles the connection and operations of the SQLite database
-class DatabaseHandler : public ocpp::common::DatabaseHandlerBase {
+class DatabaseHandler : public ocpp::common::DatabaseHandlerCommon {
 private:
     fs::path db_path;          // directory where the database file is located
```
Comment (Pietfried) [inline include/ocpp/v16/database_handler.hpp:43]: This is not used anymore and can be removed

**Review by Pietfried — CHANGES_REQUESTED** [review]
Looks nice and clean. Note some inline requested changes.

File: include/ocpp/v201/database_handler.hpp (hunk 2 above)
Comment (marcemmers) [reply include/ocpp/v201/database_handler.hpp:44]: I see a relatively easy way of doing this so will do it in this PR :+1:

File: lib/ocpp/common/database/database_handler_common.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                const std::string message_type = stmt->column_text(2);
+                const std::string message_timestamp = stmt->column_text(4);
+                const int message_attempts = stmt->column_int(3);
+
+                json json_message = json::parse(message);
+
+                DBTransactionMessage control_message;
+                control_message.message_attempts = message_attempts;
+                control_message.timestamp = message_timestamp;
+                control_message.message_type = /*conversions::string_to_messagetype(*/ message_type;
```
Comment (hikinggrass) [inline lib/ocpp/common/database/database_handler_common.cpp:48]: that comment can probably be removed

**Review by Pietfried — APPROVED** [review]
